### DISCUSSION TITLE
"init = False" do not works in relationship?

### DISCUSSION THREAD

**Discussion by charlesw1234** [sqlalchemy/sqlalchemy#9900] [Usage Questions]
I'm a greenhand of SQLAlchemy2.0 and shall use 2.0 with FastAPI, which use pydantic to rule its web API.

Firstly, my code works without pydantic:
```python
from typing import Annotated
from sqlalchemy import create_engine, MetaData, ForeignKey, String
from sqlalchemy.orm import registry as Registry, Session
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship

registry = Registry(metadata = (metadata := MetaData()))
intpk = Annotated[int, mapped_column(primary_key = True, autoincrement = True)]
class Base(DeclarativeBase):
    registry, metadata = registry, metadata

class User(Base):
    __tablename__ = 'user'
    id: Mapped[intpk]    
    name: Mapped[str] = mapped_column(String(32))
    fullname: Mapped[str | None]
    address: Mapped[list['Address']] = relationship(
        back_populates = 'user', cascade = 'all, delete-orphan')

class Address(Base):
    __tablename__ = 'address'
    id: Mapped[intpk]
    email_address: Mapped[str]
    user_id: Mapped[int] = mapped_column(ForeignKey('user.id'))
    user: Mapped['User'] = relationship(back_populates = 'address')

engine = create_engine('postgresql+psycopg://draft:draft@localhost/draft')

def create_all(): return metadata.create_all(engine)
def insert_one():
    with Session(engine) as session, session.begin():
        user = User(name = 'tester', fullname = None)
        addresses = (Address(email_address = 'tester@tester.net'),
                     Address(email_address = 'Tester@Tester.net'))
        user.address.extend(addresses)
        session.add(user)
        session.add(*addresses)
```

This code is coming from documentation of SQLAlchemy 2.0, but after pydantic is added:
```python
from typing import Annotated
from pydantic.dataclasses import dataclass
from sqlalchemy import create_engine, MetaData, ForeignKey, String
from sqlalchemy.orm import registry as Registry, Session
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship
from sqlalchemy.orm import MappedAsDataclass

registry = Registry(metadata = (metadata := MetaData()))

intpk = Annotated[int, mapped_column(
    init = False, default = 0, primary_key = True, autoincrement = True)]
class Base(MappedAsDataclass, DeclarativeBase, dataclass_callable = dataclass):
    registry, metadata = registry, metadata

class User(Base):
    __tablename__ = 'user'
    id: Mapped[intpk]
    name: Mapped[str] = mapped_column(String(32))
    fullname: Mapped[str | None]
    address: Mapped[list['Address']] = relationship(
        back_populates = 'user', cascade = 'all, delete-orphan', init = False)

class Address(Base):
    __tablename__ = 'address'
    id: Mapped[intpk]
    email_address: Mapped[str]
    user_id: Mapped[int] = mapped_column(ForeignKey('user.id'), init = False)
    user: Mapped['User'] = relationship(init = False, back_populates = 'address')

User.__pydantic_model__.update_forward_refs()

engine = create_engine('postgresql+psycopg://draft:draft@localhost/draft')

def create_all(): return metadata.create_all(engine)
def insert_one():
    with Session(engine) as session, session.begin():
        user = User(name = 'tester', fullname = None)
        addresses = (Address(email_address = 'tester@tester.net'),
                     Address(email_address = 'Tester@Tester.net'))
        user.address.extend(addresses)
        session.add(user)
        session.add(*addresses)
```

The following error is reported:
```python
>>> import sa2x1
>>> sa2x1.create_all()
>>> sa2x1.insert_one()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/wangli/aifire/aifire.draft/sqlalchemy/sa2x1.py", line 37, in insert_one
    user = User(name = 'tester', fullname = None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 4, in __init__
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/orm/state.py", line 566, in _initialize_instance
    with util.safe_reraise():
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/orm/state.py", line 564, in _initialize_instance
    manager.original_init(*mixed[1:], **kwargs)
  File "/opt/py3.aifire/lib/python3.11/site-packages/pydantic/dataclasses.py", line 325, in new_init
    self.__pydantic_validate_values__()
  File "/opt/py3.aifire/lib/python3.11/site-packages/pydantic/dataclasses.py", line 425, in _dataclass_validate_values
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for User
address
  field required (type=value_error.missing)
>>>
```

It seems that "init = False" in "    address: Mapped[list['Address']] = relationship(
        back_populates = 'user', cascade = 'all, delete-orphan', init = False)" do not works.
What can I do to fix this? thanks very much.

**Comment by CaselIT** [comment]
Hi,

This seems an issue of pydantic. It seems id does not respect `init=False` ( only if you don't provide a default?).
Try adding `defaul_factory=list`

**Comment by charlesw1234** [comment]
Thanks a lot, the previous problem is fixed, but a new encountered:
```python
from typing import Annotated
from pydantic.dataclasses import dataclass
from sqlalchemy import create_engine, MetaData, ForeignKey, String
from sqlalchemy.orm import registry as Registry, Session
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship
from sqlalchemy.orm import MappedAsDataclass

registry = Registry(metadata = (metadata := MetaData()))

intpk = Annotated[int, mapped_column(
    init = False, default = 0, primary_key = True, autoincrement = True)]
class Base(MappedAsDataclass, DeclarativeBase, dataclass_callable = dataclass):
    registry, metadata = registry, metadata

class User(Base):
    __tablename__ = 'user'
    id: Mapped[intpk]
    name: Mapped[str] = mapped_column(String(32))
    fullname: Mapped[str | None]
    address: Mapped[list['Address']] = relationship(
        back_populates = 'user', cascade = 'all, delete-orphan',
        init = False, default_factory = list)

class Address(Base):
    __tablename__ = 'address'
    id: Mapped[intpk]
    email_address: Mapped[str]
    user_id: Mapped[int] = mapped_column(ForeignKey('user.id'), init = False, default = 0)
    user: Mapped['User'] = relationship(
        init = False, default = None, back_populates = 'address')

User.__pydantic_model__.update_forward_refs()

engine = create_engine('postgresql+psycopg://draft:123dr456ft789@localhost/draft')

def create_all(): return metadata.create_all(engine)
def insert_one():
    with Session(engine) as session, session.begin():
        user = User(name = 'tester', fullname = None)
        addresses = (Address(email_address = 'tester@tester.net'),
                     Address(email_address = 'Tester@Tester.net'))
        user.address.extend(addresses)
        session.add(user)
        session.add(*addresses)
```

and the problem is:
```python
>>> import sa2x1
>>> sa2x1.insert_one()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/wangli/aifire/aifire.draft/sqlalchemy/sa2x1.py", line 43, in insert_one
    session.add(user)
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3326, in add
    self._save_or_update_state(state)
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3353, in _save_or_update_state
    for o, m, st_, dct_ in mapper.cascade_iterator(
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/orm/mapper.py", line 3941, in cascade_iterator
    queue = deque(
            ^^^^^^
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/orm/relationships.py", line 1501, in cascade_iterator
    tuples = state.manager[self.key].impl.get_all_pending(state, dict_)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/py3.aifire/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 1701, in get_all_pending
    current = getattr(current, "_sa_adapter")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'list' object has no attribute '_sa_adapter'
>>> 

```

**Reply by CaselIT** [reply]
This is exactly the same as https://github.com/sqlalchemy/sqlalchemy/discussions/9894

TLDR: pydantic does not work with external libraries well. There is really not much that sqlalchemy can do here, since pydantic completely replaces the instance dicts

**Reply by zzzeek** [reply]
TLTLTLDR: use [SQLModel](https://sqlmodel.tiangolo.com/) if you must use pydantic with SQLAlchemy.

**Reply by CaselIT** [reply]
> TLTLTLDR: use [SQLModel](https://sqlmodel.tiangolo.com/) if you must use pydantic with SQLAlchemy.

that is trying to use the same class as the both a pydantic model and a sqlalchemy mapped class

**Comment by charlesw1234** [comment]
I have checkout the source of SQLModel and read its pyproject.toml and PKG-INFO, they tell me that SQLModel is not compatible with SQLAlchemy 2.0，and the author does not update its code for seven monthes...

So I shall put the requirementation about combine pydantic with SQLAlchemy 2.0 aside and try it in recent future... May be a patch of pydantic should be written to let it can check "init" argument...

Thanks again.

**Reply by zzzeek** [reply]
that's unfortunate they havent enabled 2.0 yet, there shouldnt be any major reason for that.

you could try posting an issue with them or try porting the important part of the logic [here](https://github.com/tiangolo/sqlmodel/blob/main/sqlmodel/main.py#L495), that's where they work around all the non-compliant things pydantic does.  there's also alternatives to pydantic.
